### PR TITLE
feat: add auto moderation support

### PR DESCRIPTION
**Pull request by ooliver1**
## Summary

Adds auto moderation support, released in <https://github.com/discord/discord-api-docs/pull/4860>.

discord/discord-api-docs#5212 if merged before this PR, will need to be implemented but I am leaving that until it is documented.
- [x] discord/discord-api-docs#5116 will need implementing
- [x] discord/discord-api-docs#5242 need to update types

### API Routes

- [x] Create Auto Moderation Rule (`http.HTTPClient.create_auto_moderation_rule`, `guild.Guild.create_auto_moderation_rule`)
- [x] Modify Auto Moderation Rule (`http.HTTPClient.modify_auto_moderation_rule`, `auto_moderation.AutoModerationRule.edit`)
- [x] Delete Auto Moderation Rule (`http.HTTPClient.delete_auto_moderation_rule`, `auto_moderation.AutoModerationRule.delete`)
- [x] List Auto Moderation Rules for Guild (`http.HTTPClient.list_guild_auto_moderation_rules`, `guild.Guild.auto_moderation_rules`)
- [x] Get Auto Moderation Rule (`http.HTTPClient.get_auto_moderation_rule`, `guild.Guild.fetch_auto_moderation_rule`)

### Gateway Events

- [x] `AUTO_MODERATION_RULE_CREATE` (`on_auto_moderation_rule_create`)
- [x] `AUTO_MODERATION_RULE_UPDATE` (`on_auto_moderation_rule_update`)
- [x] `AUTO_MODERATION_RULE_DELETE` (`on_auto_moderation_rule_delete`)
- [x] `AUTO_MODERATION_ACTION_EXECUTION` (`on_auto_moderation_action_execution`, `auto_moderation.AutoModerationActionExecution`)

### Intents

- [x] `AUTO_MODERATION_CONFIGURATION` (`flags.Intents.auto_moderation_configuration`)
- [x] `AUTO_MODERATION_EXECUTION` (`flags.Intents.auto_moderation_execution`)
- [x] Shortcut for both (`flags.Intents.auto_moderation`)

### Audit Log Entries

- [x] `AUTO_MODERATION_RULE_CREATE` (`enums.AuditLogAction.auto_moderation_rule_create`)
- [x] `AUTO_MODERATION_RULE_UPDATE` (`enums.AuditLogAction.auto_moderation_rule_update`)
- [x] `AUTO_MODERATION_RULE_DELETE` (`enums.AuditLogAction.auto_moderation_rule_delete`)
- [x] `AUTO_MODERATION_BLOCK_MESSAGE` (`enums.AuditLogAction.auto_moderation_block_message`)

### Objects

- [x] Auto Moderation Rule Object (`auto_moderation.AutoModerationRule`)
- [x] Trigger Types (`enums.AutoModerationTriggerType`)
- [x] Trigger Metadata (`auto_moderation.AutoModerationTriggerMetadata`)
- [x] Keyword Preset Types (`enums.KeywordPresetType`)
- [x] Event Types (`enums.AutoModerationEventType`)
- [x] Auto Moderation Action Object (`auto_moderation.AutoModerationAction`)
- [x] Action Types (`enums.AutoModerationActionType`)
- [x] Action Metadata (`auto_moderation.AutoModerationActionMetadata`)

### Other

- [x] Message Type `AUTO_MODERATION_ACTION` (`enums.MessageType.auto_moderation_action`)
- [x] Guild Feature `AUTO_MODERATION`

## Checklist

- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by ooliver1** [comment]
Adding another metadata field would be as easy as just editing that one metadata class. In your way, adding another field would mean
- editing attributes for rule
- editing parameters for create
- adding parameters for edit

File: docs/api.rst [hunk 1]
```diff
@@ -2465,6 +2501,85 @@ of :class:`enum.Enum`.
...
 
+    .. attribute:: auto_moderation_rule_create
+
+        An auto moderation rule was created.
+
+        WHen this is the action, the type of :attr:`~AuditLogEntry.target` is
+        the :class:`AutoModerationRule` or :class:`Object` with the ID of the
+        rule which was created.
+
+        Possile attributes for :class:`AuditLogDiff`:
```
Comment (Skelmis) [inline docs/api.rst:2512]: *Possible

File: docs/api.rst [hunk 2]
```diff
@@ -2465,6 +2501,85 @@ of :class:`enum.Enum`.
 
         .. versionadded:: 2.0
 
+    .. attribute:: auto_moderation_rule_create
+
+        An auto moderation rule was created.
+
+        WHen this is the action, the type of :attr:`~AuditLogEntry.target` is
```
Comment (Skelmis) [inline docs/api.rst:2508]: *When

File: docs/api.rst [hunk 3]
```diff
@@ -2465,6 +2501,85 @@ of :class:`enum.Enum`.
...
+        - :attr:`~AuditLogDiff.event_type`
+        - :attr:`~AuditLogDiff.name`
+        - :attr:`~AuditLogDiff.trigger_type`
+        - :attr:`~AuditLogDiff.trigger_metadata`
+
+        .. versionadded:: 2.1
+
+    .. attribute:: auto_moderation_rule_update
+
+        An auto moderation rule was update.
```
Comment (Skelmis) [inline docs/api.rst:2527]: *updated

File: docs/api.rst [hunk 4]
```diff
@@ -2465,6 +2501,85 @@ of :class:`enum.Enum`.
...
+
+        WHen this is the action, the type of :attr:`~AuditLogEntry.target` is
```
Comment (Skelmis) [inline docs/api.rst:2529]: *When

File: docs/api.rst [hunk 5]
```diff
@@ -2465,6 +2501,85 @@ of :class:`enum.Enum`.
...
+        - :attr:`~AuditLogDiff.trigger_type`
+        - :attr:`~AuditLogDiff.trigger_metadata`
+
+        .. versionadded:: 2.1
+
+    .. attribute:: auto_moderation_rule_deleted
+
+        An auto moderation rule was deleted.
+
+        WHen this is the action, the type of :attr:`~AuditLogEntry.target` is
```
Comment (Skelmis) [inline docs/api.rst:2550]: *When

File: nextcord/http.py [hunk 6]
```diff
@@ -2337,3 +2338,88 @@ def get_event_users(
...
+        return self.request(r)
+
+    def create_auto_moderation_rule(
+        self,
+        guild_id: Snowflake,
+        data: auto_moderation.AutoModerationRuleCreate,
+        *,
+        reason: Optional[str] = None,
+    ) -> Response[auto_moderation.AutoModerationRule]:
+        valid_keys = (
```
Comment (Skelmis) [inline nextcord/http.py:2368]: Could we put both of these `valid_keys` into one location?

File: nextcord/guild.py [hunk 7]
```diff
@@ -3463,3 +3469,148 @@ async def register_application_commands(self, *commands: BaseApplicationCommand)
...
+        Forbidden
+            You do not have permission to create auto moderation rules.
+        HTTPException
+            Creating the rule failed.
+        InvalidArgument
+            An invalid type was passed for an argument.
+
+        Returns
+        -------
+        :class:`AUtoModerationRule`
```
Comment (Skelmis) [inline nextcord/guild.py:3579]: *AutoModerationRule

**Review by Skelmis — CHANGES_REQUESTED** [review]
Havent viewed auto_moderation yet.

File: nextcord/http.py (hunk 6 above)
Comment (ooliver1) [reply nextcord/http.py:2368]: `trigger_type` is not included in `modify_`

**Comment by ooliver1** [comment]
Edited description to note about https://github.com/discord/discord-api-docs/pull/5212, this will have to be watched to decide if this PR gets it or this will be merged before that docs PR.

File: docs/api.rst [hunk 8]
```diff
@@ -3387,6 +3575,52 @@ AuditLogDiff
...
+
+    .. attribute:: event_type
+
+        The event type of an auto moderation rule being changed.
+
+        :type: :class:`AutoModerationEventType`
+
+    .. attribute:: actions
+
+        The list of actions of an auto moderation rule being changed.
```
Comment (Skelmis) [inline docs/api.rst:3598]: The list of actions being changed.

File: nextcord/guild.py [hunk 9]
```diff
@@ -3463,3 +3469,148 @@ async def register_application_commands(self, *commands: BaseApplicationCommand)
...
+        data = await self._state.http.get_auto_moderation_rule(self.id, rule_id)
+        return AutoModerationRule(data=data, state=self._state)
+
+    async def create_auto_moderation_rule(
+        self,
+        *,
+        name: str,
+        event_type: AutoModerationEventType,
+        trigger_type: AutoModerationTriggerType,
+        actions: List[AutoModerationAction],
```
Comment (Skelmis) [inline nextcord/guild.py:3534]: this is undoccumented

File: nextcord/auto_moderation.py [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+    __slots__ = ("keyword_filter", "presets")
+
+    def __init__(
+        self,
+        keyword_filter: Optional[List[str]] = None,
+        presets: Optional[List[KeywordPresetType]] = None,
+    ) -> None:
+        self.keyword_filter = keyword_filter
+        self.presets = presets
+
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:94]: Want to explicitly type these?

File: nextcord/auto_moderation.py [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+        .. note::
+
+            The maximum value that can be used is `2419200` seconds (4 weeks)
+    """
+
+    __slots__ = ("channel_id", "duration_seconds")
+
+    def __init__(self, channel: Optional[Snowflake] = None, duration_seconds: Optional[int] = None):
+        self.channel_id = channel.id if channel is not None else None
+        self.duration_seconds = duration_seconds
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:150]: Explicitly type these?

File: nextcord/auto_moderation.py [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+            :attr:`AutoModerationActionType.timeout`.
+
+        .. note::
+
+            The maximum value that can be used is `2419200` seconds (4 weeks)
+    """
+
+    __slots__ = ("channel_id", "duration_seconds")
+
+    def __init__(self, channel: Optional[Snowflake] = None, duration_seconds: Optional[int] = None):
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:148]:
```
def __init__(self, *, channel: Optional[Snowflake] = None, duration_seconds: Optional[int] = None):
```

File: nextcord/auto_moderation.py [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+            :attr:`AutoModerationTriggerType.keyword_preset`.
+    """
+
+    __slots__ = ("keyword_filter", "presets")
+
+    def __init__(
+        self,
+        keyword_filter: Optional[List[str]] = None,
+        presets: Optional[List[KeywordPresetType]] = None,
+    ) -> None:
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:91]: Enforce keyword argument only here

File: nextcord/auto_moderation.py [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+
+    __slots__ = ("type", "metadata")
+
+    def __init__(
+        self,
+        type: AutoModerationActionType,
+        metadata: Optional[AutoModerationActionMetadata] = None,
+    ) -> None:
+        self.type = type
+        self.metadata = metadata
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:201]: Explicitly type these

File: nextcord/auto_moderation.py [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+            payload["channel_id"] = self.channel_id
+
+        if self.duration_seconds is not None:
+            payload["duration_seconds"] = self.duration_seconds
+
+        return payload
+
+
+class AutoModerationAction:
+    """Represents an auto moderation action which will execute whenever a rule is triggered.
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:174]: Represents an auto moderation action that will execute whenever a rule is triggered.

File: nextcord/auto_moderation.py [hunk 16]
```diff
@@ -0,0 +1,503 @@
...
+        The rule's unique ID.
+    guild_id: :class:`int`
+        The guild's unique ID which this rule belongs to.
+    guild: Optional[:class:`Guild`]
+        The guild which this rule belongs to, if found in cache.
+    name: :class:`str`
+        The rule's name.
+    creator_id: :class:`int`
+        The user's unique ID which first created this rule.
+    creator: :class:`Member`
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:256]: This should be `creator: Optional[:class:`Member`]` right?

File: nextcord/auto_moderation.py [hunk 17]
```diff
@@ -0,0 +1,503 @@
...
+        self.exempt_roles = (
+            [self.guild.get_role(role_id) for role_id in self.exempt_role_ids]
+            if self.guild is not None
+            else []
+        )
+        self.exempt_channels = (
+            [self.guild.get_channel(channel_id) for channel_id in self.exempt_channel_ids]
+            if self.guild is not None
+            else []
+        )
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:321]: Typehint these

File: nextcord/auto_moderation.py [hunk 18]
```diff
@@ -0,0 +1,503 @@
...
+        The ID of the system alert message, if sent.
+    alert_system_message: Optional[:class:`Message`]
+        The system alert message, if sent and found in cache.
+    action: :class:`AutoModerationAction`
+        The action that was executed.
+    rule_id: :class:`int`
+        The id of the rule that was executed.
+    rule_trigger_type: :class:`AutoModerationTriggerType`
+        The type of rule that was executed.
+    user_id: :class:`int`
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:464]: Should this param not be `member_id` given its in a guild?

File: nextcord/auto_moderation.py [hunk 19]
```diff
@@ -0,0 +1,503 @@
...
+        The ID of the user that triggered this action.
+    user: Optional[:class:`User`]
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:466]: See prior

File: nextcord/auto_moderation.py [hunk 20]
```diff
@@ -0,0 +1,503 @@
...
+        self.alert_system_message_id = _get_as_snowflake(data, "alert_system_message_id")
+        self.alert_system_message = state._get_message(self.alert_system_message_id)
+        self.action = AutoModerationAction.from_data(data=data["action"])
+        self.rule_id = int(data["rule_id"])
+        self.rule_trigger_type = try_enum(AutoModerationTriggerType, data["rule_trigger_type"])
+        self.user_id = int(data["user_id"])
+        self.user = state.get_user(self.user_id)
+        self.content = data["content"]
+        self.matched_keyword = data["matched_keyword"]
+        self.matched_content = data["matched_content"]
```
Comment (Skelmis) [inline nextcord/auto_moderation.py:503]: typehint these

**Review by Skelmis — CHANGES_REQUESTED** [review]

File: docs/api.rst [hunk 21]
```diff
@@ -2465,6 +2501,85 @@ of :class:`enum.Enum`.
...
+        - :attr:`~AuditLogDiff.trigger_metadata`
+
+        .. versionadded:: 2.1
+
+    .. attribute:: auto_moderation_block_message
+
+        A message was blocked by an auto moderation rule.
+
+        When this is the action, the type of :attr:`~AuditLogEntry.target` is
+        the :class:`Member` or :class:`User` who's message was blocked.
```
Comment (DenverCoder1) [inline docs/api.rst:2572]:
```suggestion
        the :class:`Member` or :class:`User` whose message was blocked.
```

**Review by DenverCoder1 — COMMENTED** [review]
Looks good, I haven't tested

**Review by DenverCoder1 — APPROVED** [review]
Looks good assuming it is tested

**Review by Skelmis — APPROVED** [review]
Untested but code looks good to me

**Pull request merged**